nemo/collections/nlp/models/language_modeling/megatron_t5_prompt_learning_model.py

```diff
@@ -0,0 +1,350 @@
+
+    def validation_epoch_end(self, outputs):
+        averaged_loss, val_preds, val_acc = self.inference_epoch_end(outputs)
+
+        # we can only log on one rank if it is rank zero so we broadcast from last rank
+        torch.distributed.broadcast(averaged_loss, get_last_rank())
+        self.log('val_loss', averaged_loss, prog_bar=True, rank_zero_only=True)
+
+        self.log('val_acc', val_acc, prog_bar=True)
+        logging.info(f'Validation loss: {averaged_loss}')
+        logging.info(f'Validation accuracy: {val_acc}')
+
+    def test_step(self, batch, batch_idx):
+        return self.validation_step(batch, batch_idx)
+
+    def test_epoch_end(self, outputs):
+        averaged_loss = average_losses_across_data_parallel_group(outputs)
+        logging.info(f'test_loss: {averaged_loss[0]}')
+
+    def build_virtual_prompt_dataset(
+        self, dataset_paths, batch_size, for_train, drop_last, shuffle, num_workers, pin_memory
+    ):
+        dataset = T5PromptLearningDataset(
+            datasets=dataset_paths,
+            tokenizer=self.tokenizer,
+            virtual_prompt_source=self.virtual_prompt_source,
+            task_templates=self.task_templates,
+            pseudo_tokens=self.pseudo_tokens,
+            pad_token_id=self.pad_token_id,
+            max_seq_length=self.cfg.data.get('max_seq_length', self.frozen_model.cfg.max_position_embeddings),
+            min_seq_length=self.cfg.data.get('min_seq_length', 1),
+            add_bos=self.cfg.data.get('add_bos', False),
+            add_eos=self.cfg.data.get('add_eos', True),
+            for_train=for_train,
+        )
+
+        rank = parallel_state.get_data_parallel_rank()
+        world_size = parallel_state.get_data_parallel_world_size()
+        sampler = torch.utils.data.distributed.DistributedSampler(
+            dataset, num_replicas=world_size, rank=rank, shuffle=shuffle
+        )
+
+        dataloader = torch.utils.data.DataLoader(
+            dataset,
+            collate_fn=dataset.collate_fn,
+            sampler=sampler,
+            batch_size=batch_size,
+            drop_last=drop_last,
+            num_workers=num_workers,
+            pin_memory=pin_memory,
+        )
+        print('build success', len(dataloader), dataset_paths)
+        return dataset, dataloader
+
+    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> Any:
+
+        enc_input, dec_input, labels, loss_mask, enc_mask, dec_mask, position_ids, taskname_ids = batch
+
+        input_embeds = self.embed_input_inference(enc_input, taskname_ids)
+
+        position_embeddings = self.frozen_model.enc_dec_model.encoder_embedding.position_embeddings(position_ids)
+        encoder_input = input_embeds + position_embeddings
+
+        predicted_token_ids, log_probs = self.frozen_model.decode(
+            tokens_enc=enc_input,
+            enc_mask=enc_mask,
+            num_tokens_to_generate=self.decoder_seq_length,
+            encoder_input=encoder_input,
```

Comment:
I think there is an issue on the main branch right now wherein `encoder_input` will just be ignored even if you pass it. I think @michalivne is working on a fix. Lets communicate with him about this.

Reply:
Hi @michalivne ! I slightly change the args in `get_forward_output_only_func` so that `encoder_input` can be utilized. I see you are working on a fix. Please let me know when your fix is done and merged.

Reply:
@michalivne wanted to follow up on this. We are shooting to merge in Jiaqi's PR before the code freeze next week. Any update on when your PR will be merged?